### PR TITLE
Use ruff linter and formatter

### PR DESCRIPTION
**Pull request by jl-wynen**
First step for https://github.com/scipp/copier_template/issues/70

I adapted the setup from Scitacean because I'm happy with it. But please check whether the selected rules are a good choice.

I did not enable `D` because there are too many violations. But it allows us to enforce a docstring format and can help avoid rst pitfalls. I'm in favour of enabling it, but please comment!

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -38,18 +28,14 @@ repos:
...
-    rev: 1.7.4
-    hooks:
-      - id: bandit
-        additional_dependencies: ["bandit[toml]"]
-        args: ["-c", "pyproject.toml"]
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.3.7
+    hooks:
+      - id: ruff-format
+        types_or: [ python, pyi ]
```
Comment (MridulS) [inline .pre-commit-config.yaml:35]: I think we can run the formatter on jupyter notebooks too.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -38,18 +28,14 @@ repos:
...
+      - id: ruff
+        args: [ --fix, --exit-non-zero-on-fix ]
```
Comment (MridulS) [inline .pre-commit-config.yaml:37]: I don't think `--exit-non-zero-on-fix` is really required here, pre-commit should catch any changed files.

**Comment by YooSunYoung** [comment]
> I did not enable D because there are too many violations. But it allows us to enforce a docstring format and can help avoid rst pitfalls. I'm in favour of enabling it, but please comment!

I'm up for enabling docstring formatting!

File: pyproject.toml [hunk 3]
```diff
@@ -108,15 +99,47 @@ skips = [
...
+ignore = [
+    "E111", "E114", "E117", "D206", "D300",  # conflict with ruff format
+]
+fixable = ["I001", "B010"]
+isort.known-first-party = ["scipp"]
+pydocstyle.convention = "numpy"
+
+[tool.ruff.lint.per-file-ignores]
+"src/scipp/__init__.py" = ["I"]
+"src/scipp/core/__init__.py" = ["I"]
```
Comment (MridulS) [inline pyproject.toml:121]:
```suggestion
"__init__.py" = ["I"]
```

should ignore in all `__init__.py` files.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (jl-wynen) [reply .pre-commit-config.yaml:37]: You are right. And the docs even recommend running the linter before the formatter. Changed it now

File: pyproject.toml (hunk 3 above)
Comment (jl-wynen) [reply pyproject.toml:121]: Why? Most `__init__.py` files are simple and can be safely reordered.

**Comment by jl-wynen** [comment]
I tried out the pyupgrade fixes. There are some useful ones that we should probably do. But we'd have to run it again when we drop python 3.9. Since that happens soon, I'd just leave it for now and run pyupgrade later.

File: pyproject.toml (hunk 3 above)
Comment (MridulS) [reply pyproject.toml:121]: I think it's just a good practice and we can generalize this upstream then, sorting `__init__.py` is usually a major source of random import errors.

File: pyproject.toml [hunk 4]
```diff
@@ -108,15 +99,47 @@ skips = [
...
+"tools/*" = [
+    "S101",  # asserts are fine, we do not run tools with optimizations
+    "T201",  # printing is ok for internal tools
+]
+"*.ipynb" = [
+    "E501",  # longer lines are sometimes more readable
+    "I",  # we don't collect imports at the top
+    "S101",  # asserts are used for demonstration and are safe in notebooks
+    "T201",  # printing is ok for demonstration purposes
+]
```
Comment (SimonHeybrock) [inline pyproject.toml:137]: 👍 for putting detailed reasons!

**Comment by MridulS** [comment]
Shouldn't https://github.com/scipp/scipp/blob/main/setup.cfg also be removed? All the config now is in pyproject.toml

**Review by nvaytet — APPROVED** [review]
Let's go ahead with this :+1:

**Comment by jl-wynen** [comment]
I added the remaining rules from https://github.com/scipp/esssans/pull/127. Some required a lot of fixes, esp `PT`, but I think it is good to have a consistent style.

I found a broken test thanks to `PT011`. :-)

**Review by MridulS — APPROVED** [review]

**Pull request merged**